src/main/scala/sbtcrossproject/CrossProject.scala

```diff
@@ -1,184 +1,119 @@
-    mapProjects(_.enablePlugins(ps: _*))
+  private def copy[T <: Platform](
+      projects: Map[Platform, Project] = projects): CrossProject[T] =
+    new CrossProject(id, root, projects)
 
-  def enablePlugins(ns: Plugins*): CrossProject =
-    mapProjects(_.enablePlugins(ns: _*))
+  def disablePlugins(ps: AutoPlugin*): CrossProject[A] =
+    copy(projects.map((x) => x._1 -> x._2.disablePlugins(ps: _*)))
 
-  def in(dir: File): CrossProject =
-    mapCrossPlatforms(p => projects(p).in(crossType.dir(p, dir)))
+  def enablePlugins(ns: Plugins*): CrossProject[A] =
+    copy(projects.map((x) => x._1 -> x._2.enablePlugins(ns: _*)))
 
-  def overrideConfigs(cs: Configuration*): CrossProject =
-    mapProjects(_.overrideConfigs(cs: _*))
+  def overrideConfigs(cs: Configuration*): CrossProject[A] =
+    copy(projects.map((x) => x._1 -> x._2.overrideConfigs(cs: _*)))
 
-  /** Configures how settings from other sources, such as .sbt files, are
-   *  appended to the explicitly specified settings for this project.
-   *
-   *  Note: If you disable AutoPlugins here, non-JVM platforms will not work
-   */
-  def settingSets(select: AddSettings*): CrossProject =
-    mapProjects(_.settingSets(select: _*))
+  def dependsOn(deps: CrossClasspathDependency*): CrossProject[A] =
+    copy(
+        projects.map((x) => x._1 -> x._2.dependsOn(deps.map(_.dep(x._1)): _*)))
 
-  def settings(ss: Def.Setting[_]*): CrossProject =
-    mapProjects(_.settings(ss: _*))
+  def crossType(ct: CrossType): CrossProject[A] = {
+    copy(ct.modify(root, projects))
+  }
 
-  override def toString(): String = {
+  override def toString: String = {
     val entries = this.projects.map {
-      case (platform, project) => "$platform = $project"
+      case (t, p) => s"${t.name} -> $p"
     }
-    s"CrossProject(${entries.mkString(", ")})"
+    s"CrossProject(${entries.mkString(",\n")})"
   }
-
-  // Helpers
-
-  private def mapCrossPlatforms(transform: CrossPlatform => Project): CrossProject =
-    copy(projects.map { case (p, _) => p -> transform(p) })
-
-  private def mapProjects(transform: Project => Project): CrossProject =
-    copy(projects.mapValues(transform))
-
-  private def copy(projects: Map[CrossPlatform, Project] = projects): CrossProject =
-    new CrossProject(crossType, projects)
 }
 
 object CrossProject extends CrossProjectExtra {
-  def apply(id: String, platforms: Seq[CrossPlatform], base: File,
-      crossType: CrossType): CrossProject =
-    apply(platforms.map(p => (p, id + p.name)).toMap, base, crossType)
-
-  def apply(ids: Map[CrossPlatform, String], base: File,
-      crossType: CrossType): CrossProject = {
 
-    val sss = sharedSrcSettings(crossType)
-
-    val projects = ids.map { case (p, id) =>
-      val project = Project(id, crossType.dir(p, base)).
-        settings(sss: _*).
-        enablePlugins(p.plugin.toSeq: _*)
-
-      p -> project
+  def crossProject_impl[A <: Platform](c: Context): c.Expr[Builder[A]] = {
```

Comment:
This does not need to be polymorphic. It can always return `CrossProject[Platform]`.